File: winning_player.rb
class Player
  attr_reader :score

  @@all_players = []

  def initialize(name, score)
    @name, @score = name, score
  end

  def self.winner_player(all_players)
    compare = ->(p1, p2) { p1.score > p2.score ? p1 : p2 }
    all_players.reduce(&compare)
  end

  def self.best_score(all_players)
    all_players.map(&:score).max
  end

  def self.all_players
    @@all_players
  end

  def self.add_player(player)
    @@all_players << player
  end
end

Player.add_player Player.new('Peter', 7.2)
Player.add_player Player.new('Steve', 8.2)
Player.add_player Player.new('Bruce', 9.2)

score = player.best_score(player.all_players)
puts "best score: #{score}"

File: winning_player.js
// The problem
//
// Get the player with greater score
var players = [
  {name: 'Steve', score: 6.2},
  {name: 'Bruce', score: 7.2},
  {name: 'Peter', score: 5.4},
];

function compareScore(p1, p2) {
  return (p1.score > p2.score) ? p1 : p2;
}

// Imperative style
var greater = players[0];
for (var i = 1; i < players.length; i++) {
  greater = compareScore(players[i], greater);
}
console.log( greater );

// Functional style
var winningPlayer = players.reduce(compareScore);
console.log( winningPlayer );
